views/default/js/elgg/upgrader.js

```diff
@@ -0,0 +1,239 @@
+				$(json.system_messages.error).each(function(key, message) {
+					if ($.inArray(message, errorMessages) === -1) {
+						var msg = '<li>' + message + '</li>';
+						messageList.append(msg);
+						messages.push(message);
+					}
+				});
+			}
+
+			$(json.output.errors).each(function(key, message) {
+				var msg = '<li>' + message + '</li>';
+				messageList.append(msg);
+				messages.push(message);
+			});
+
+			numSuccess = parseInt(json.output.numSuccess);
+			numError = parseInt(json.output.errors.length);
+
+			numProcessed += (numSuccess + numError);
+
+			// Increase success statistics
+			counter.text(numProcessed + '/' + total);
+
+			// Increase the progress bar
+			progressbar.progressbar({ value: numProcessed });
+
+			if (numError > 0) {
+				errorCounter
+					.text(elgg.echo('upgrade:error_count', [messages.length]))
+					.css('color', 'red');
+			}
+
+			updateCounter();
+
+			var percentage = 0;
+			if (numProcessed < total) {
+				percentage = parseInt(numProcessed * 100 / total);
+
+				// Increase percentage
+				percent.html(percentage + '%');
+
+				// Start next upgrade call
+				processBatch();
+			} else {
+				if (numError > 0) {
+					// Upgrade finished with errors. Give instructions on how to proceed.
+					elgg.register_error(elgg.echo('upgrade:finished_with_errors'));
+				}
+
+				// Increase percentage
+				percent.html('100%');
+
+				// Reset all counters
+				numSuccess = numError = numProcessed = percentage = 0;
+				messages = [];
+
+				// Get next upgrade
+				upgrade = upgrade.next();
+
+				if (upgrade.length) {
+					// Continue to next upgrade
+					runUpgrade();
+				} else {
+					spinner.stop();
+					$('#upgrade-finished').removeClass('hidden');
+				}
+			}
+		};
+
+		// We use post() instead of action() so we can catch error messages
+		// and display them manually underneath the upgrade view.
+		return elgg.post('action/admin/upgrade', options);
+	};
+
+	/**
+	 * Displays estimated amount of time needed for a single upgrade to finish
+	 */
+	function updateCounter () {
+		var now = new Date().getTime();
+
+		// How many milliseconds ago the last batch was started
+		var difference = (now - upgradeStartTime) / 1000;
+
+		// How many items are waiting to be processed
+		var unProcessed = total - numProcessed;
+
+		var timeLeft = Math.round((difference / numProcessed) * unProcessed);
+
+		if (timeLeft < 60) {
+			var hours = '00';
+			var minutes = '00';
+			var seconds = timeLeft;
+		} else {
+			if (timeLeft < 3600) {
+				var minutes = Math.floor(timeLeft / 60);
+				var seconds = timeLeft % 60;
+				var hours = '00';
+			} else {
+				var hours = Math.floor(timeLeft / 3600);
+				var timeLeft = timeLeft % 3600;
+				var minutes = Math.floor(timeLeft / 60);
+				var seconds = timeLeft % 60;
+			}
+		}
+
+		hours = formatDigits(hours);
+		minutes = formatDigits(minutes);
+		seconds = formatDigits(seconds);
+
+		var value = hours + ':' + minutes + ':' + seconds;
+
+		timer.html(value);
+	};
```

Comment:
This should probably go to its own AMD module. Perhaps nowadays there's even some existing library that could take care of the job. (I didn't find one back when I made the code.)